runtime/v2/README.md

```diff
@@ -1,29 +1,177 @@
+this will be the runtime _shim_. Either way, this is the binary that containerd executes and expects to start the ttRPC listener.
+The runtime name can be either a URI-like string, or, beginning with containerd 1.6.0, the actual path to the executable.
+
+1. If the runtime name is a path, use that as the actual path to the runtime to invoke.
+1. If the runtime name is URI-like, convert it to a runtime name using the below logic.
+
+If the runtime name is URI-like, containerd will convert the passed runtime from the URI-like name to a binary name using the following logic:
+
+1. Replaces all `.` with `-`
+1. Takes the last 2 components, e.g. `runc.v2`
+1. Prepends `containerd-shim`
+
+For example, if the runtime name is `io.containerd.runc.v2`, containerd will invoke the shim as `containerd-shim-runc-v2`. It expects to
+find the binary in its normal `PATH`.
+
+containerd keeps the `containerd-shim-*` prefix so that users can `ps aux | grep containerd-shim` to see running shims on their system.
+
+For example:
+
 ```bash
-> ctr run --runtime io.containerd.runc.v2
+$ ctr --runtime io.containerd.runc.v2 run --rm docker.io/library/alpine:latest alpine
 ```
 
-When a user specifies a runtime name, `io.containerd.runc.v2`, they will specify the name and version of the runtime.
-This will be translated by containerd into a binary name for the shim.
+Will invoke `containerd-shim-runc-v2`.
 
-`io.containerd.runc.v2` -> `containerd-shim-runc-v2`
-
-Since 1.6 release, it's also possible to specify absolute runtime path:
+You can test this by trying another name:
 
 ```bash
-> ctr run --runtime /usr/local/bin/containerd-shim-runc-v2
+$ ctr run --runtime=io.foo.bar.runc2.v2.baz --rm docker.io/library/hello-world:latest hello-world /hello
+ctr: failed to start shim: failed to resolve runtime path: runtime "io.foo.bar.runc2.v2.baz" binary not installed "containerd-shim-v2-baz": file does not exist: unknown
 ```
 
-containerd keeps the `containerd-shim-*` prefix so that users can `ps aux | grep containerd-shim` to see running shims on their system.
+It received `io.foo.bar.runc2.v2.baz` and looked for `containerd-shim-v2-baz`.
+
+You also can override the default configured runtime for the shim, by passing it the `--runc-binary`
+option. For example"
+
+```
+ctr --runtime io.containerd.runc.v2 --runc-binary /usr/local/bin/runc-custom run --rm docker.io/library/alpine:latest alpine
+```
+
+### Configuring Runtimes
+
+You can configure multiple runtimes in containerd's `config.toml` configuration file, by modifying the
+section:
+
+```toml
+      [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
+```
+
+See [config.toml man page](../../docs/man/containerd-config.toml.5.md) for more details and an example.
+
+#### Using configured runtimes
+
+**TODO:** Since we select runtimes by passing both the runtime to invoke (normally a shim) and the binary name (e.g. `runc`) - either via gRPC
```

Comment:
I think we've now answered this in the other discussion.  The config file is used for CRI, which does not allow the user to specify a shim or options directly.

Reply:
I don't find it so clear, even after the above, at least from the docs. I imagine someone coming at it fresh will be confused.

> The config file is used for CRI, which does not allow the user to specify a shim or options directly.

So in CRI, I can specify only a name, or nothing at all (i.e. default), and no options?

If so:

1. When can I specify specific runtime path/URI and options? Via gRPC? I assume that `ctr` uses that API?
2. Why is the CRI called in the config file `plugins."io.containerd.grpc.v1.cri".containerd.runtimes`?

Reply:
> So in CRI, I can specify only a name, or nothing at all (i.e. default), and no options?

The CRI proto defines a [`runtime_handler` field](https://github.com/kubernetes/cri-api/blob/de5f1318aede866435308f39cb432618a15f104e/pkg/apis/runtime/v1/api.proto#L476) which is either blank or populated with some name.  If that name maps to a runtime configuration that is defined in the containerd config file, then containerd will use that defined config.

> When can I specify specific runtime path/URI and options? Via gRPC? I assume that ctr uses that API?

Yes, `ctr` and `nerdctl` use the containerd thick client library (which is currently the files at the top-level of this repository) and under the hood that speaks containerd's built-in gRPC API.  `ctr` and `nerdctl` do not speak CRI.

> Why is the CRI called in the config file `plugins."io.containerd.grpc.v1.cri".containerd.runtimes`?

I'm not sure I'm understanding this question.

CRI is a specific gRPC API defined by [this proto definition](https://github.com/kubernetes/cri-api/blob/de5f1318aede866435308f39cb432618a15f104e/pkg/apis/runtime/v1/api.proto).  The CRI plugin in containerd implements the interface defined in that file and serves it over containerd's socket.  The CRI plugin accepts configuration via the config file in sections named `plugins."io.containerd.grpc.v1.cri"`.

Reply:
OK, got it. Removed that and made the last but a bit clearer. I also squashed the two commits.

Anything else?